front/pages/api/data_sources/[user]/[name]/search.ts

```diff
@@ -0,0 +1,167 @@
+import { NextApiRequest, NextApiResponse } from "next";
+import { User, DataSource, Provider } from "@app/lib/models";
+import { credentialsFromProviders } from "@app/lib/providers";
+import { parse_payload } from "@app/lib/http_utils";
+import { JSONSchemaType } from "ajv";
+import { auth_user } from "@app/lib/auth";
+
+type TagsFilter = {
+  in?: string[];
+  not?: string[];
+};
+
+type TimestampFilter = {
+  gt?: number;
+  lt?: number;
+};
+
+type SearchFilter = {
+  tags?: TagsFilter;
+  timestamp?: TimestampFilter;
+};
+
+export type DatasourceSearchQuery = {
+  query: string;
+  top_k: number;
+  full_text: boolean;
+  timestamp_gt?: number;
+  timestamp_lt?: number;
+  tags_in?: string[];
+  tags_not?: string[];
+};
+
+const searchQuerySchema: JSONSchemaType<DatasourceSearchQuery> = {
+  type: "object",
+  properties: {
+    query: { type: "string" },
+    top_k: { type: "number" },
+    full_text: { type: "boolean" },
+    timestamp_gt: { type: "number", nullable: true },
+    timestamp_lt: { type: "number", nullable: true },
+    tags_in: { type: "array", items: { type: "string" }, nullable: true },
+    tags_not: { type: "array", items: { type: "string" }, nullable: true },
+  },
+  required: ["query", "top_k", "full_text"],
+};
+
+const { DUST_API } = process.env;
+
+type DatasourceSearchResponseBody = {
+  documents: Array<Document>;
+};
+
+export default async function handler(
+  req: NextApiRequest,
+  res: NextApiResponse<DatasourceSearchResponseBody>
+): Promise<void> {
+  const [authRes, dataSourceUser] = await Promise.all([
+    auth_user(req, res),
+    User.findOne({
+      where: {
+        username: req.query.user,
+      },
+    }),
+  ]);
+
+  if (authRes.isErr()) {
+    res.status(authRes.error().status_code).end();
+    return;
+  }
+  const auth = authRes.value();
+
+  if (!dataSourceUser) {
+    res.status(404).end();
+    return;
+  }
+
+  const dataSource = await DataSource.findOne({
+    where: {
+      userId: dataSourceUser.id,
+      name: req.query.name,
+    },
+  });
+
+  if (!dataSource) {
+    res.status(404).end();
+    return;
+  }
+
+  switch (req.method) {
+    case "GET": {
+      if (!auth.canEditDataSource(dataSource)) {
```

Comment:
We may want to add a small comment explaining why `canEditDataSource` here and not `canReadDataSource`?